### PR TITLE
Assemble, Linking BC, LLD Link in ROCm

### PR DESCRIPTION
**Pull request by raikonenfnu**
-Generate object + Link with bitcode
-ld.lld link object to generate code object
-Added Target Chip and Link Bitcode Option for ROCm

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Target/ROCM/ROCMTarget.cpp [hunk 1]
```diff
@@ -144,12 +146,18 @@ class ROCMTargetBackend final : public TargetBackend {
     iree_compiler::FlatbufferBuilder builder;
     iree_ROCMExecutableDef_start_as_root(builder);
 
+    // Link module to Device Library
+#ifdef LINK_ROCM_ON
```
Comment (ThomasRaoux) [inline iree/compiler/Dialect/HAL/Target/ROCM/ROCMTarget.cpp:150]: I would just use a `llvm::cl::opt<bool>` as it doesn't need to be a compile time choice. This way it makes the build simpler and allow to decide at runtime whether to link or not with command line.

Comment (raikonenfnu) [reply iree/compiler/Dialect/HAL/Target/ROCM/ROCMTarget.cpp]: Done :)

File: iree/compiler/Dialect/HAL/Target/ROCM/ROCMTargetUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+#include "iree/compiler/Dialect/HAL/Target/ROCM/ROCMTarget.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+#include "mlir/Support/LogicalResult.h"
+
+// MLIR Support Files Header
```
Comment (ThomasRaoux) [inline iree/compiler/Dialect/HAL/Target/ROCM/ROCMTargetUtils.cpp:23]: nit: comment doesn't seem completely accurate or very useful :). I would remove it.

**Review by ThomasRaoux — APPROVED** [review]

File: iree/compiler/Dialect/HAL/Target/ROCM/ROCMTargetUtils.cpp (hunk 2 above)
Comment (raikonenfnu) [reply iree/compiler/Dialect/HAL/Target/ROCM/ROCMTargetUtils.cpp]: Done, I'll ping you again, once all the test pass :)

Comment (ThomasRaoux) [reply iree/compiler/Dialect/HAL/Target/ROCM/ROCMTargetUtils.cpp]: Thanks! I enabled auto-merge once everything passes.

**Pull request merged**